### ISSUE TITLE
SQL Server Primary key contrainst index with options DATA_COMPRESSION

### ISSUE THREAD

**Issue by deadcativan** [dotnet/efcore#33145]
In #30408 index options "DATA_COMPRESSION" is supported, I think there is another place where the same thing is needed. It's the primary key when creating a table for an entity.

Currently I create a custom class extends SqlServerAnnotationProvider to add annotation for data compression as following:

```
    public class ExtendedSqlServerAnnotationProvider : SqlServerAnnotationProvider
    {
        public ExtendedSqlServerAnnotationProvider(RelationalAnnotationProviderDependencies dependencies) : base(dependencies)
        {
        }

        public override IEnumerable<IAnnotation> For(IUniqueConstraint constraint, bool designTime)
        {
            foreach (var annotation in base.For(constraint, designTime))
            {
                yield return annotation;
            }

            if (!designTime)
            {
                yield break;
            }

            var key = constraint.MappedKeys.First();

            var table = constraint.Table;

            var dataCompression = key.GetDataCompression(StoreObjectIdentifier.Table(table.Name, table.Schema));
            if(dataCompression.HasValue)
            {
                yield return new Annotation(SqlServerAnnotationNames.DataCompression, dataCompression.Value);
            }

        }
    }
```

And then I extend SqlServerMigrationsSqlGenerator to build SQL for index options according to this annotation in AddPrimaryKeyOperation and AddUniqueConstraintOperation:

**Comment by roji** [comment]
Makes sense - @hmajerus are you interested in submitting a PR for this, as you did the original PR for the index case (#30408)?

**Comment by hmajerus** [comment]
> Makes sense - @hmajerus are you interested in submitting a PR for this, as you did the original PR for the index case (#30408)?

Yeah, I can look into it later next week.

**Comment by hmajerus** [comment]
So, did not get to this last week. Fairly busy lately and for the next month so won't get to this anytime soon, but think I could do it for the EF9 release. If someone else wants to do it before then that's fine with me.

**Comment by m-x-shokhzod** [comment]
Opened #38434 for this.

It adds `UseDataCompression` on `KeyBuilder` / `KeyBuilder<TEntity>` / `IConventionKeyBuilder` (plus `CanSetDataCompression`) and `Get`/`SetDataCompression` on the key metadata, mirroring the existing key `FillFactor` and index `DATA_COMPRESSION` support. `SqlServerAnnotationProvider.For(IUniqueConstraint)` attaches the annotation, and since `PrimaryKeyConstraint`/`UniqueConstraint` already route through the shared `IndexOptions` (which already emits `DATA_COMPRESSION`), the migrations generator itself is unchanged. As with the index case, it is not reverse-engineered.
